Documentation/llilc-at-six-months.md

```diff
@@ -0,0 +1,456 @@
+It wasn�t that bad because we already had to demote structures into memory 
+at call boundaries.
+
+### Explicit Exception Checks
+
+Many MSIL opcodes can cause semantically meaningful exceptions (some of them
+more than one kind). We were accustomed to IR that could support operation and 
+operand exceptions � for instance loads and stores via null pointers could raise
+`NullReference` exceptions � with the actual exception check being implicit in
+the IR. LLVM IR, however, has an exception model that matches up better with
+C++, where exceptions can only happen at calls. While we (and others) in the
+LLVM community have had discussions about enabling somewhat more general models
+in LLVM IR, for now we make all our exception checks explicit in the IR.
+
+The current LLVM direction seems like a reasonably good compromise for an AOT 
+compiler: explicit checks in HIR with the ability to fold them into implicit
+checks at the machine level. For a JIT, this is something we�ll potentially
+revisit down the road, though the issue of which model is preferable is far
+from clear, with tradeoffs on a number of fronts. 
+
+Implicit checks may not be efficient on all architectures. In high-level IR,
+the ability for a wide variety of operations to raise exceptions imposes a
+semantic burden on IR maintenance and optimization. For instance, hoisting a
+load that can throw now has control flow implications.
+
+Explicit checks give the IR a nice �separation of concerns� aspect � a load or
+store won�t have complex behaviors. The check IR has the same shape and
+semantic as other compare and branch operations and so is more naturally
+amenable to common optimizations. But there is extra IR volume created via the
+explicit checks, and this impacts the throughput and code size of a JIT. For
+example, when compiling `Buffer.Memmove`, LLILC ends up generating 85 explicit
+null checks. LLILC places the throw blocks out of line, and so the tail end of
+the method looks like this:
+
+```LLVM
+ThrowNullRef:                                     ; preds = %20
+  call void bitcast (i64* @"AnyJITHelper::JitHelper" to void ()*)() #1, !dbg !22
+  unreachable, !dbg !22
+
+ThrowNullRef2:                                    ; preds = %23
+  call void bitcast (i64* @"AnyJITHelper::JitHelper" to void ()*)() #1
+  unreachable
+
+ThrowNullRef4:                                    ; preds = %28
+  call void bitcast (i64* @"AnyJITHelper::JitHelper" to void ()*)() #1, !dbg !23
+  unreachable, !dbg !23
+
+; ... (82 more)...
+```
+
+Despite having identical instruction sequences, these blocks can�t be tail
+merged because they carry unique debugging information, and this information is
+crucial to preserve, since it allows users to diagnose which check actually
+failed. To be fair, many of these checks can be optimized away, but that puts
+us in a bit of a pickle, since we might now require fairly throughput-intensive
+optimizations to get decent code size and code quality.
```

Comment:
FWIW, I had to re-read this a few times to grok what you were getting at -- my initial response was "wait, isn't it expected that you might need optimizations to get decent code size and code quality?".  I think the pickle that you're wanting to emphasize is not so much about "decent code size and code quality" as it is about "reasonable IR size (i.e. compact enough to allow reasonable throughput)" -- the pickle is that we may need throughput-intensive optimizations to get throughput-friendly IR.  Unless I'm missing the point?